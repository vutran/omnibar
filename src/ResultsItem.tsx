import * as React from 'react';
import { COLORS, DEFAULT_HEIGHT } from './constants';
import { AnchorItem } from './modifiers/anchor';
import AnchorRenderer from './modifiers/anchor/AnchorRenderer';

interface Props<T> {
  // results renderer function
  children: Omnibar.ResultRenderer<T>;
  // the item
  item: T;
  // onMouseEnter item callback
  onMouseEnter?: (e: any /* Event */) => void;
  // onMouseLeave item callback
  onMouseLeave?: (e: any /* Event */) => void;
  // onClick callback
  onClickItem?: (e: any /* Event */) => void;
  // set to true to highlight the given item
  highlighted?: boolean;
  // optional style override
  style?: React.CSSProperties;
}

interface State {
  // set to true to highlight
  hover: boolean;
}

const ITEM_STYLE: React.CSSProperties = {
  height: DEFAULT_HEIGHT,
  lineHeight: `${DEFAULT_HEIGHT}px`,
  fontSize: 24,
  borderStyle: 'solid',
  borderColor: COLORS.DARKGRAY,
  borderTopWidth: 0,
  borderLeftWidth: 1,
  borderRightWidth: 1,
  borderBottomWidth: 1,
  boxSizing: 'border-box',
};

const ITEM_HOVER_STYLE: React.CSSProperties = {
  backgroundColor: COLORS.GRAY,
};

export default class ResultRenderer<T> extends React.PureComponent<
  Props<T>,
  State
> {
  static defaultProps = {
    highlighted: false,
  };

  state: State = {
    hover: false,
  };

  handleMouseEnter = (evt: any /* Event */) => {
    this.setState({ hover: true });
    this.props.onMouseEnter && this.props.onMouseEnter(evt);
  };

  handleMouseLeave = (evt: any /* Event */) => {
    this.setState({ hover: false });
    this.props.onMouseLeave && this.props.onMouseLeave(evt);
  };

  render() {
    const item = this.props.item;
    let style: React.CSSProperties = { ...ITEM_STYLE, ...this.props.style };

    if (this.props.highlighted || this.state.hover) {
      style = { ...style, ...ITEM_HOVER_STYLE };
    }

    const renderer = this.props.children
      ? this.props.children
      : (AnchorRenderer as Omnibar.ResultRenderer<T>);

    return renderer({
      style,
      item,
      onMouseEnter: this.handleMouseEnter,
      onMouseLeave: this.handleMouseLeave,
      onClick: this.props.onClickItem,
    });
  }
}
