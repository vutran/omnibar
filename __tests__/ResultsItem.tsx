import * as React from 'react';
import ResultsItem from '../src/ResultsItem';
import renderer from 'react-test-renderer';

describe('ResultsItem', () => {
  it('should render a <ResultsItem />', () => {
    const item = { title: 'Google', url: 'https://google.com' };
    const tree = renderer
      .create(React.createElement(ResultsItem, { item, children: null }))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render a highlighted <ResultsItem />', () => {
    const item = { title: 'Google', url: 'https://google.com' };
    const tree = renderer
      .create(
        React.createElement(ResultsItem, {
          item,
          isSelected: true,
          children: null,
        })
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
