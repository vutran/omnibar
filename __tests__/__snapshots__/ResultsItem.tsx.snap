// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultsItem should render a <ResultsItem /> 1`] = `
<a
  href="https://google.com"
  item={
    Object {
      "title": "Google",
      "url": "https://google.com",
    }
  }
  onClick={undefined}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "borderBottomWidth": 1,
      "borderColor": "#ddd",
      "borderLeftWidth": 1,
      "borderRightWidth": 1,
      "borderStyle": "solid",
      "borderTopWidth": 0,
      "boxSizing": "border-box",
      "color": "#000",
      "display": "block",
      "fontSize": 24,
      "height": 50,
      "lineHeight": "50px",
      "paddingLeft": 15,
      "paddingRight": 15,
      "textDecoration": "none",
    }
  }
>
  Google
</a>
`;

exports[`ResultsItem should render a highlighted <ResultsItem /> 1`] = `
<a
  href="https://google.com"
  item={
    Object {
      "title": "Google",
      "url": "https://google.com",
    }
  }
  onClick={undefined}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "backgroundColor": "#eee",
      "borderBottomWidth": 1,
      "borderColor": "#ddd",
      "borderLeftWidth": 1,
      "borderRightWidth": 1,
      "borderStyle": "solid",
      "borderTopWidth": 0,
      "boxSizing": "border-box",
      "color": "#000",
      "display": "block",
      "fontSize": 24,
      "height": 50,
      "lineHeight": "50px",
      "paddingLeft": 15,
      "paddingRight": 15,
      "textDecoration": "none",
    }
  }
>
  Google
</a>
`;
